Sources/Core/Client.swift

```diff
@@ -623,91 +698,112 @@ public actor Client {
+                self.attachmentMap[docKey]?.peerPresenceMap[publisher] = presence
 
-                    self.attachmentMap[key]?.peerPresenceMap[publisher] = presence
-                case .UNRECOGNIZED:
-                    break
-                }
-            }
+                self.sendPeerChangeEvent(.watched, [docKey], publisher)
+            case .documentsUnwatched:
+                self.sendPeerChangeEvent(.unwatched, [docKey], publisher)
 
-            switch pbWatchEvent.type {
+                self.attachmentMap[docKey]?.peerPresenceMap.removeValue(forKey: publisher)
             case .documentsChanged:
-                let event = DocumentsChangedEvent(value: responseKeys)
-                self.eventStream.send(event)
-            case .documentsWatched, .documentsUnwatched, .presenceChanged:
-                let event = PeerChangedEvent(value: keys.reduce([String: [String: Presence]](), self.getPeersWithDocKey(peersMap:key:)))
+                self.attachmentMap[docKey]?.remoteChangeEventReceived = true
+
+                let event = DocumentsChangedEvent(value: [docKey])
                 self.eventStream.send(event)
+            case .presenceChanged:
+                if let peerPresence = self.attachmentMap[docKey]?.peerPresenceMap[publisher], peerPresence.clock > presence.clock {
+                    break
+                }
+
+                self.attachmentMap[docKey]?.peerPresenceMap[publisher] = presence
+
+                self.sendPeerChangeEvent(.presenceChanged, [docKey], publisher)
             case .UNRECOGNIZED:
                 break
             }
         }
     }
 
-    private func onStreamDisconnect() {
-        self.watchLoopTask?.cancel()
-        self.watchLoopTask = nil
+    private func sendPeerChangeEvent(_ type: PeersChangedValue.`Type`, _ keys: [DocumentKey], _ actorID: ActorID? = nil) {
+        let value = PeersChangedValue(type: type, peers: keys.reduce([DocumentKey: PresenceMap]()) {
+            self.getPeersWithDocKey(peersMap: $0, docKey: $1, actorID: actorID)
+        })
+        let event = PeerChangedEvent(value: value)
 
-        self.watchLoopReconnectTimer = Timer(timeInterval: Double(self.reconnectStreamDelay) / 1000, repeats: false) { _ in
-            Task {
-                await self.doWatchLoop()
-            }
+        self.eventStream.send(event)
+    }
+
+    private func disconnectWatchStream(_ docKey: DocumentKey) throws {
+        guard self.attachmentMap[docKey] != nil else {
+            throw YorkieError.documentNotAttached(message: "\(docKey) is not attached.")
         }
 
-        if let watchLoopReconnectTimer = self.watchLoopReconnectTimer {
-            RunLoop.main.add(watchLoopReconnectTimer, forMode: .common)
+        guard self.attachmentMap[docKey]?.remoteWatchStream != nil else {
+            return
         }
 
+        self.attachmentMap[docKey]?.remoteWatchStream?.cancel()
+        self.attachmentMap[docKey]?.remoteWatchStream = nil
+
```

Comment:
Should we also need to invalidate watchLoopReconnectTimer to prevent re-connecting the stream when detaching?

Reply:
I'll add the codes for invalidating that timer.